Refactor to Typescript

beverts312 opened the pull request:
This PR is to address #52 

A few things to note/discuss...

I had to adjust these from rules on the generator from the rules the generator generated (the rules outputed by the generator are still the same):
- no-any (index.ts: 23&30)
- missing-jsdoc
- max-func-body-length - bumped from 25 to 30
- linebreak-style - previous setting only allowed windows  

Additionally I added inline exceptions for:  
- Line length on a few names
- Function names for index. (so we could allow `_functions`)  

Any problems with these adjustments?

Additionally, the current version of the yeoman typedefinitions are bad...I have checked in a corrected version to this branch, I will open a pr on DefinitelyTyped so we can get it fixed

calebcartwright commented on .eslintignore:
```diff
@@ -1,3 +1,6 @@
 coverage
-testresults
\ No newline at end of file
+testresults
```
I might be turned around, but is this configuration reversed? when we run eslint against the generator, we do want it to hit the transpiled index.js, tests, etc. I was originally hoping we could configure things such that when we would lint the generator codebase, we could also get it to lint the template files, but I kept encountering failures when it did that because the template files have variables/substitution tokens/etc. that make the linters error out

calebcartwright commented on .gitignore:
```diff
@@ -1,5 +1,5 @@
 # NPM/Node
-/node_modules/
+node_modules/*
```
Just adding our discussion from earlier. If we can package up the typedef into our own module (or even a GitHub repo) and do an npm install instead of having it & other content checked into source then I think that would be better.

All depends on the amount of effort for that vs. getting that PR into the definitelytyped repo (assuming they publish to npmjs fairly quickly)

beverts312 replied:
Can we merge as is and track the [PR](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/14850) with an issue?

calebcartwright commented on build/istanbul-config.js:
```diff
@@ -1,24 +1,20 @@
-var path = require('path');
+var unitTestCoverageDirectory = './coverage/unit';
+var unitTestCoverageReportHtmlFile = './coverage/unit/index.html';
 
```
why remove the path.resolve on the coverage directory and the html file? I can't remember the specifics, but I had to add these to resolve an issue

beverts312 replied:
I replaced the build directory with the ones generator by our generator, so I think this is more a question for you ;)

calebcartwright replied:
Lol! I do that so much.. As I imagine will be the case with other things pulled from the generator, it will require a fix in two places (the updated source for the generator code itself, and the template content it came from)

calebcartwright commented on build/istanbul-config.js:
```diff
@@ -1,24 +1,20 @@
 
-var unitTestCoverageDirectory = path.resolve('./coverage/unit/');
-var unitTestCoverageReportHtmlFile = path.resolve('./coverage/unit/index.html');
-
```
Was the thinking here that the comments were redundant? I put them in there because I was previously unaware that we could apply thresholds to each file as well as at an aggregate, but if the var names are explanatory enough then fair enough

calebcartwright replied:
I am guessing this came from the template content which is missing the comments

calebcartwright commented on build/mocha-config.js:
```diff
@@ -1,7 +1,7 @@
 
 module.exports = {
-    unitTestTimeout: 5000,
+    unitTestTimeout: 3000,
```
I had to bump this up due to some timeout issues on one of the build tools (forget which). Honestly it should be like.. 50ms but just need to keep an eye on it if things start failing again

calebcartwright replied:
If this came from the template content, can we update them both back to 5000?

beverts312 replied:
yep

calebcartwright commented on build/tasks/test.js:
```diff
@@ -59,7 +58,7 @@ gulp.task('enforce-code-coverage', ['run-unit-tests'], function() {
 });
 
-gulp.task('show-unit-test-coverage-report', ['run-unit-tests'], function() {
+gulp.task('show-unittest-coverage-report', ['run-unit-tests'], function() {
```
Thank you for making this consistent with the npm script name :) can't tell you how many times I have mixed the two up

calebcartwright commented on generators/app/index.ts:
```diff
@@ -152,38 +165,40 @@ module.exports = yeoman.Base.extend({
-        var context = this._buildExpressContext();
+        const context = this._buildExpressContext();
 
-        this.fs.copyTpl(glob.sync(this.sourceRoot() + '/**/*', { dot: true }), this.destinationRoot(), context);
```
is the glob sync not necessary here anymore? are the dot files/directories still included?

beverts312 replied:
This caused a tslint error, according to the typedefs glob.sync outputed an array of strings and copyTpl only took a string, I did have to make 2 adjustments for the dotfiles (copying what you had to do for some)

calebcartwright replied:
weird. either there is an issue with the type def or copyTpl is smart enough to handle an array. as long as we've got the updates for all the other 'dot' files then all good

calebcartwright commented on generators/app/index.ts:
```diff
@@ -1,78 +1,40 @@
+const expressRoot = path.join(templateRoot, 'express-api');
+const vscodeRoot = path.join(templateRoot, 'vscode');
+
+class Generator extends yeoman.Base {
```
thoughts on SwellabyGenerator for the classname?

beverts312 replied:
much better

calebcartwright commented on jsconfig.json:
```diff
@@ -1,6 +0,0 @@
-{
```
Can we leave this file in here? VS Code uses it (or at least used to) for JS, and we will still have some JS files (gulp tasks, etc.)

beverts312 replied:
sure

calebcartwright commented on package.json:
```diff
@@ -45,14 +45,16 @@
   },
   "scripts": {
-    "build": "npm run lint check-security run-unit-tests enforce-code-coverage",
+    "build": "gulp clean tslint transpile",
```
I had trouble getting Travis-CI to execute the original script (that invoked other npm scripts) but can we try adding eslint, enforce-code-coverage, check-security, etc. so that build tools can be updated to do a simple: npm run build?

beverts312 replied:
yep

calebcartwright commented on temp:
```diff
@@ -0,0 +1,155 @@
+    _validateDirectoryName: () => {
```
What is this file?

beverts312 replied:
woops lol

calebcartwright commented on test/unit/app/index-tests.ts:
```diff
@@ -31,143 +28,145 @@ var boilerplateFiles = [
+                    });
+            });
+        // tslint:disable-next-line:max-line-length
+        test('Should scaffold into the current directory when the specified app name matches the current directory name with the Base option',
```
What about breaking the string up into multiple lines? That's what I've typically done w/long test names

beverts312 replied:
At my resolution this is more readable ;) Either way is fine with me (I assume it would work fine). Which way do you want?

calebcartwright commented on test/unit/app/input-config-tests.js:
```diff
@@ -1,281 +0,0 @@
-'use strict';
```
Is the difference in # of lines here the removal of the tests around the "required" attribute you were telling me about earlier?

calebcartwright replied:
Or is this another instance where I missed something in the template content

beverts312 replied:
I think about 60

calebcartwright commented:
general comment to add to the record here: there's been a lot of stuff I have missed/fat-fingered/or otherwise got wrong in the template content, so a lot of the things commented in probably stem from issues in the templates that have been around for a while. 

this is a good exercise as besides the TS experience, we are also taking care of template issues :)

beverts312 commented:
Going to dump everything we have left on this here (easier to track than above):
- Jsdoc error is coming from tslint
- Moved the no-any and max-func-length to in-line exceptions
- Reverted changes caused by generator & made those changes to the generated code
- Moved more stuff into the `npm run build` but we clean & transpile twice, do you want to take a stab at that?
- Opened a PR to definitely typed for the yeoman stuff (linked above), I propose we keep it as is here for now and track the PR with an issue, thoughts?
- I havnt messed with the test line lengths yet, you want me to disable or fix?

calebcartwright commented:
aligning with the most recent bullets above:

- disabled in TS is fine, as long as we enable the intelligent rule in ESLint in the near future
- excellent
- double excellent
- gulp shouldn't be doing that as long as we have our prereq tasks defined correctly, but will take a look
- I think we are good to merge the PR in prior to the definitelytyped PR, but one last question on this note.. I know it wouldn't make any substantive difference (apart from making me feel better about gitignore file) but would it still work if we had the index.d.ts file checked in somewhere else (not an npm install) until defintelytyped is fixed?
- I vote for breaking the lines up. Open for a discussion around what the threshold should be, but I feel like that rule exists for precisely these types of scenarios (it is also cut in half on GitHub, for me at least)

beverts312 commented:
everything is done except double transpile still happens on build, we can merge and open a bug for you for that? or you can work of this branch? doesnt matter to me

calebcartwright commented:
go ahead and merge, I will take a look at the double transpile later on and will open an issue if needed

beverts312 commented:
to make this thread easier to read I deleted all but the last report from coveralls
